Add MKL ndarray and MKL library check for the new MKL engine

pcs-theano opened the pull request:
As previously mentioned in issue #5580, we would like to merge our optimization with MKL engine into the master branch. Since there're a lot of code change, so we separate our implementations into several different PRs for the upstream. 

This is our first PR for the optimization, we created a new folder `mkl` under sandbox for all the new codes. Also we added a new tensor type named `MKLNdarrayType` for the data with MKL layout and buffer. The new data type will be used for all the MKL related Ops.

The rest of the code changes will be separated into different PRs based on each different Op and its optimizer in the following days after this one.

Any comments are welcome.

pcs-theano commented:
Currently, there're two kinds of mkl library which supports dnn functionality.
1) mkl library after version `2017.0.0` (build date: `20160801)`, which could be installed with parallel studio 2017 in common way.
2) mklml which is a small version of mkl, and it's free and easy to get and install. You can find the latest mklml from [here](https://github.com/01org/mkl-dnn/releases/download/v0.7/mklml_lnx_2017.0.2.20170209.tgz)

In order to make it more convenient to use, we're planning to submit another PR to add a script for downloading and setting up environment for mklml library in theano, which could be run independently or via setup.py as below:
	`python setup.py install --mkl`
	
Any comments or suggestions?

nouiz commented on theano/__init__.py:
```diff
@@ -126,6 +126,12 @@ def test():
+if config.device.startswith('cpu'):
+    try:
+        import theano.sandbox.mkl
+    except Exception as e:
```
This is a very broad catch and could hide syntax error and other such things. Can you be more strict in what you catch?

pcs-theano replied:
Sure, we can handle with the known exceptions here, while leave others as a warning instead of a simple "pass" to be more meaningful.

nouiz commented on theano/sandbox/mkl/__init__.py:
```diff
@@ -0,0 +1,254 @@
+import textwrap
+import stat
+import shutil
+import theano
```
nitpik, add an empty line before the import of Theano to separate the different import group. (we aren't always consistent in Theano, but we try to go in that direction, at one point, we would probably do it in all files)

pcs-theano replied:
No problem, I will change it in this way to catch that point.

nouiz commented on theano/sandbox/mkl/__init__.py:
```diff
@@ -0,0 +1,254 @@
+                                theano.Mode(optimizer=None),
+                                profile=False)
+            mkl_version.v = f()
+        except:
```
Why this try except? If there is an error here, is that a sign of a problem we want to raise to the user?

If this need to stay, you should always catch some exception, otherwise, it will also catch interruption like ctrl-C and we don't want that. Be the most precise in which exception you catch (Exception is better then not setting one. But ValueError is better then Exception).

pcs-theano replied:
Sure, I will add "Exception" here for the case of failing to compile or link with MKL library.

nouiz commented on theano/sandbox/mkl/__init__.py:
```diff
@@ -0,0 +1,254 @@
+    if config.device != "cpu":
+        mkl_available.msg = "MKL is disabled since device is not CPU. " \
+                            "Set the device to 'CPU' if you want to use MKL."
+        return False
```
If config.mkl.nn.enabled is False, you should raise an error.

pcs-theano replied:
config.mkl.nn.enabled=False allows user not to use MKL even if mkl is available. If raising an error here, there's only one way to go. 
Not flexible?

nouiz replied:
Sorry, my first comment should have been:

If config.mkl.nn.enabled is True, you should raise an error.

(True instead of False)

pcs-theano replied:
Sure, will raise an error when device is not CPU but mkl.nn.enabled is True, otherwise will return False.

nouiz commented on theano/sandbox/mkl/__init__.py:
```diff
@@ -0,0 +1,254 @@
+            try_run=True, output=True, compiler=theano.config.cxx,
+            comp_args=False)
+
+        mkl_available.msg = (
```
I think you should set this message only if compilation_ok isn't True or run_ok isn't True

pcs-theano replied:
OK, I will add the condition here.

nouiz commented on theano/sandbox/mkl/__init__.py:
```diff
@@ -0,0 +1,254 @@
+
+                    from mkl_ndarray.mkl_ndarray import *  # noqa
+                except Exception as e:
+                    _logger.error('Failed to compile mkl_ndarray.c: %s', str(e))
```
All that remind me that we want to refactor/ease compiling other lib with Theano. It probably won't be done shortly, so don't change the PR, but we should do it at some point.

pcs-theano replied:
Currently, mkl_ndarray is developed and validated only with Theano, so we follow the similar way as cuda. Maybe we can build such package as a third party lib to Theano in future.

nouiz replied:
I'm not sure my comment was clear. I mean, I don't have problems to have that inside Theano. The problem we have is that it is hard to do so in Theano, and we do that for a few things already (the old gpu back-end, cutils, scan cython version, in pylearn2 for cuda convnet (the old alex net cuda code)). My only point is that we should make this easier for Theano to compile lib that aren't Theano op.

pcs-theano replied:
Got it, sounds great, we can keep the current code as it is, and align to the new design when it's available.

nouiz commented on theano/sandbox/mkl/__init__.py:
```diff
@@ -0,0 +1,254 @@
+# register name of 'mkl_opt' in opt.py and then add tags for it.
+try:
+    from . import opt
+    opt.optdb.add_tags('mkl_opt', 'fast_compile', 'fast_run')
```
Do that call only if we want to enable mkl part. that should not be done if mkl_available() return False.

I think you can just increate the indentation of this try block by one tab

pcs-theano replied:
I will remove the try and except and move the other codes within `if mkl_available()`

nouiz commented on theano/sandbox/mkl/__init__.py:
```diff
@@ -0,0 +1,254 @@
+try:
+    from . import opt
+    opt.optdb.add_tags('mkl_opt', 'fast_compile', 'fast_run')
+except:
```
I think you can remove the try except here comptly. This will put the add_tags inside "if mkl_available()" and at that point, it should not fail.

nouiz commented on theano/sandbox/mkl/mkl_ndarray.h:
```diff
@@ -0,0 +1,86 @@
+    dnnLayout_t private_layout;                 // layout instance create by MKL APIs
+    void*       private_data;                   // data buffer
+    void*       private_workspace;              // computation workspace for forward and backward
+}MKLNdarray;
```
Why not reuse numpy ndarray obejct?

nouiz commented on theano/sandbox/mkl/mkl_ndarray.h:
```diff
@@ -0,0 +1,86 @@
+__attribute__((visibility ("default"))) const size_t* MKLNdarray_DIMS(const MKLNdarray* self);
+__attribute__((visibility ("default"))) const size_t* MKLNdarray_STRIDES(const MKLNdarray* self);
+__attribute__((visibility ("default"))) int MKLNdarray_NDIM(const MKLNdarray* self);
+__attribute__((visibility ("default"))) int MKLNdarray_TYPE(const MKLNdarray* self);
```
Use a macro to define the visibility as is done with the MOD_PUBLIC macro. Maybe you can reuse that one.

pcs-theano replied:
Will change the code in that way.

nouiz commented on theano/sandbox/mkl/mkl_type.py:
```diff
@@ -0,0 +1,270 @@
+    pass
+
+
+class MKLNdarrayType(Type):
```
Before doing a full review of that, why do you want a new C object and not reuse numpy ndarray object or our libgpuarray object. libgpuarray object isn't just for GPU.

If there is a good reason to have a new C object defined in Theano, then I think we need to refactor that, as there is a big quantity of copied code from TensorType. You could make a GenericTensorType() object that both of them inherit from it, and maybe GpuArrayType (but we could do this in another PR).

pcs-theano replied:
Let me introduce some basic background about Intel MKL DNN primitives firstly.
 
Intel MKL DNN primitives work with special opaque data types such as operation primitives (such as convolution forward/backward primitive, etc.,) and layouts of processed data (specifies relative location of elements of processed arrays in memory). Each DNN operation requires input and output arrays have certain data layouts.
 
More detail information can be found [here](https://software.intel.com/en-us/node/684759)
 
 
The reasons why we added such a new C object are: 
 
1.	Numpy ndarray and GPU ndarray doesn’t support storing extra variable such as layout structure.
The memory view described by Numpy ndarray and GPU ndarray is via dims and strides, while it’s described by an opaque data types (layouts) in MKL.
Besides, layouts needs to be passed from one Op to the next one for the computation. We need an C object (TensorType) which allows us to pass the layout.

2.	The data buffer of Numpy ndarray and GPU ndarray can’t be reused by MKL due to different size.
The size of data buffer of Numpy ndarray is exact same as described by dims and strides. While the size of data buffer of mkl op is described by its layout, and it might have different size compared with numpy (might include padding)
 
3.	Same as item 1.
For some Ops, backward pass needs the intermediate result which calculated in forward pass. So we need another data buffer named "workspace" in the variable.
 
Actually, in the beginning, we're thinking about leveraging Numpy ndarray or Cuda/GPU ndarray for this purpose, but none of them allows us to do so. So that's why we create a simplified but sufficient structure for mkl Ops. 

We're glad to see If there's any better suggestions.

nouiz replied:
I'm not sure I get everything. By layout, what do you mean? I see 2 interpretations that could be used here that are close to each others and used in Theano.
1) The memory layout. specificed at 100% by shapes and strides. It don't give any semantic to each dimensions, just how to read the data in the right order.
2) The "semantic layout". Is the layout for the inputs images (batch, in channel, image height, image width) or the other semantic layout as tensorflow, cuda convnet and nervana use different layout.

In one, do you mean that your C object have the "semantic layout" inside it?

I checked rapidly the URL you gave. A small comments. I saw many fct that ends with F32 or F64. I suppose it mean float32 or float64. Can you confirm?

pcs-theano replied:
I mean the first one, layout specifies relative location of elements in memory. 
Different MKL DNN operation (for example, convolution, pool, etc.,) might use different layout , and even if the same operation, it could have different layout in different layers or topologies according to their parameters (such as kernel/windows, stride, padding).

You're correct about F32/F64, there're two sets of functions for float32 and float64 currently, and more precision will be supported in future.

nouiz replied:
I understand your last explanation. But I don't understand why this is a reason for a new C object. shapes and strides in the data object are enough to specify the memory layout of your point 1) 2 comments above.

nouiz commented on theano/sandbox/mkl/mkl_type.py:
```diff
@@ -0,0 +1,270 @@
+    ndim = None
+    broadcastable = None
+
+    def __init__(self, broadcastable, name=None, dtype=None):
```
This is the order from our old gpu-back-end. It had a mistake in the order compared to the CPU. Please, use the order 

    def __init__(self, dtype, broadcastable, context_name=None, name=None):

If we end up needing a new Type.

nouiz commented on theano/sandbox/mkl/mkl_type.py:
```diff
@@ -0,0 +1,270 @@
+    def clone(self, dtype=None, broadcastable=None):
+        if broadcastable is None:
+            broadcastable = self.broadcastable
+        return self.__class__(broadcastable, name=self.name, dtype=dtype)
```
```
if dtype is None:
    dtype = self.dtype
```

nouiz commented on theano/sandbox/mkl/mkl_type.py:
```diff
@@ -0,0 +1,270 @@
+
+    def filter(self, data, strict=False, allow_downcast=None):
+        # return self.filter_inplace(data, None, strict=strict, allow_downcast=allow_downcast)
+        return data
```
The goal of filter is to cast data into a data represented by this type. Currently it don't do that.

nouiz commented on theano/sandbox/mkl/mkl_ndarray.c:
```diff
@@ -0,0 +1,1278 @@
+
+        memset((void*)(self->user_structure), 0, 2 * MAX_NDIM * sizeof (size_t));
+    } else {
+        PyErr_SetString(PyExc_RuntimeError, "MKLNdarray_new: fail to create a new instance \n");
```
Should this be an PyExc_MemoryError?

nouiz commented on theano/sandbox/mkl/mkl_ndarray.c:
```diff
@@ -0,0 +1,1278 @@
+    // copy data to mkl buffer
+    size_t element_size = (size_t)PyArray_ITEMSIZE(obj);
+    // assert (user_size * element_size <= self->data_size);
+    memcpy((void*)self->private_data, (void*)PyArray_DATA(obj), user_size * element_size);
```
What if the ndarray is strided? Currently it don't raise an error nor copy correctly in that case.

pcs-theano commented:
@nouiz 

The MKL layout concept is essential just memory format description, it not only contains shapes, strides but some other stuff which related with hardware specifications. So we can't leverage the current Numpy ndarrays's structure to represent it.

All MKL DNN primitives use their own layouts, all these layouts are optimized based on the tensors' logical size and what the current hardware is (for example, Xeon, Xeon Phi have different layout for the same Numpy ndarray), in order to show good efficiency for DNN Ops. It's much complicated than NCHW or NHWC, etc., which are used by current DL FWs.

And the buffer allocated based on MKL layout includes rearranged data, and might includes some extra information. So the memory size might be larger than Numpy ndarray's data buffer.

Since the layout structure needs to be passed from one Op to the next, and a bigger buffer is required to store the optimal rearrange data, that's why we need a special C object to support it. 

The good thing is all this stuff are no matter for end users, since the MKL Ops handles all of them. 

Currently, there's no any page for MKL layout. Hope the above description could help you better understand it.

nouiz commented:
I think I got part of why a new object is needed. If I combine the information from a few places, as you sometimes use a 6d inputs tensor where some of the normal dimensions are split into 2 non concecutive dimensions. In that case, you need to track not just how to read the data from the RAM, but also the meaning of each dimensions. And that request something. Using a new C object for that make sence.

@lamblin suggested to create a new directory called "contrib" and put your code there instead of sandbox, what do you think of that?

As we understand the why of the new object, you can do the first batch of comments I did. After that I'll have another look at it.

thanks

pcs-theano commented:
@nouiz 

Thanks for your time to review and discuss about the new C object, glad to see that it's accepted, it will make the other MKL Ops easy since they're all built based on this new object.

We'll update the PR according to your previous comments, and move all our codes from sandbox to "contrib".

BTW, regarding the script which supports automatically downloading the small MKL library as I mentioned in the very beginning of this issue page, any comments for that?

nouiz commented:
I like the idea to help user install mkl. So if you have a script or can add that to setup.py, it would be great. I don't have a preference. Using setup.py could make this easier to install on windows then scripts.

nouiz commented on theano/contrib/mkl/mkl_ndarray.h:
```diff
@@ -0,0 +1,91 @@
+#define MKL_FLOAT32  (11)
+#define MKL_FLOAT64  (12)
+
+char* MKL_TYPE[] = {"", "", "", "int16", "", "int32", "", "int64",
```
does some ints are supported?

pcs-theano replied:
Not yet, but they should be supported in future.

nouiz commented on theano/contrib/mkl/mkl_type.py:
```diff
@@ -0,0 +1,350 @@
+        self.broadcastable = tuple(bool(b) for b in broadcastable)
+        self.name = name
+        self.dtype_specs()
+        if not (context_name is None):
```
Why a context_name? Is it always "mkl"?

Is this for Xeon Phi?

pcs-theano replied:
No difference for Xeon or Xeon Phi, so we remove it.

nouiz commented on theano/contrib/mkl/mkl_type.py:
```diff
@@ -0,0 +1,350 @@
+        if not (context_name is None):
+            self.context_name = context_name
+
+    def clone(self, dtype=None, broadcastable=None):
```
If we need to keep the context_name, I would add it here too for completness. But if you know it isn't useful, don't add it.

pcs-theano replied:
Ok, we just remove it.

nouiz commented on theano/contrib/mkl/mkl_type.py:
```diff
@@ -0,0 +1,350 @@
+
+    @staticmethod
+    def values_eq(a, b):
+        raise NotImplementedError('MKLNdarray values_eq')
```
This probably need to be implemented. values_eq_approx() must be implemented to be able to work in DebugMode. At worst, convert both object to ndarray and call TensorType.values_eq_approx() on the converted version.

pcs-theano replied:
Both functions were added in the new code.

nouiz commented on theano/contrib/mkl/mkl_type.py:
```diff
@@ -0,0 +1,350 @@
+        """
+        return hash((type(self), self.dtype, self.broadcastable, self.context_name))
+
+    ndim = property(lambda self: len(self.broadcastable), doc='number of dimensions')
```
Move that at the top of this class to replace your ndim = None there.

pcs-theano replied:
Done.

nouiz commented on theano/contrib/mkl/mkl_type.py:
```diff
@@ -0,0 +1,350 @@
+    """,
+    version=(1,))
+
+theano.compile.register_shape_i_c_code(
```
ideally, theano.compile.register_deep_copy_op_c_code() and 
theano.compile.register_specify_shape_c_code() should have c code as in tensor/type.py, but this isn't mandatory.

pcs-theano replied:
Both of them were added in the new codes.

pcs-theano commented:
@nouiz 
Sorry for the late response on this issue, we've updated the codes according to your comments, please help to have a check on the new changes.

pcs-theano commented on theano/contrib/mkl/__init__.py:
```diff
@@ -0,0 +1,256 @@
+
+    def f(local_opt):
+        name = (kwargs and kwargs.pop('name')) or local_opt.__name__
+        mkl_optimizer.register(name, local_opt, 'fast_run', 'fast_compile',
```
Sure, will keep it for now and align to the new change in future.

pcs-theano commented on theano/contrib/mkl/mkl_ndarray.c:
```diff
@@ -0,0 +1,1361 @@
+    // nchw -> whcn
+    for (int i = 0; i < self->nd; i++) {
+        mkl_size[i] = (MKLNdarray_DIMS(self))[self->nd - i - 1];
+        mkl_stride[i] = (MKLNdarray_STRIDES(self))[self->nd - i -1];
```
MKL object has its own description for the layout, and it's enough to MKL object.
The reason to keep the original information (strides and dims) is that, when we convert MKL object to Numpy Ndarray we need to know the layout information of target object (Numpy Ndarray).

pcs-theano commented on theano/contrib/mkl/mkl_ndarray.c:
```diff
@@ -0,0 +1,1361 @@
+        (PyCFunction)MKLNdarray_CreateArrayObj, METH_VARARGS,
+        "Copy from MKL to a numpy ndarray."},
+
+    {"zeros",
```
Both of them were added in the new codes.

pcs-theano commented on theano/contrib/mkl/mkl_ndarray.c:
```diff
@@ -0,0 +1,1361 @@
+        }
+
+        if (self->private_workspace) {
+            rval = dnnReleaseBuffer_F32(self->private_workspace);
```
Regarding the dnn function names, it was a legacy design in MKL that including dtype in it. 

So with such fixed APIs, we can't avoid duplicated codes, no matter we write them in this way or create a new header file which combines the two functions into a wrapper by using a parameter "dtype" to distinguish different data types.

So I suggest to keep the code as it was for now, and align to MKL when it evolves.

pcs-theano commented on theano/contrib/mkl/mkl_ndarray.c:
```diff
@@ -0,0 +1,1361 @@
+    npy_intp* d = PyArray_DIMS(obj);
+    int typenum = PyArray_TYPE(obj);
+
+    if (typenum != MKL_FLOAT32 && typenum != MKL_FLOAT64) {
```
Code was updated according to your comments.

pcs-theano commented on theano/contrib/mkl/mkl_ndarray.c:
```diff
@@ -0,0 +1,1361 @@
+
+    int err = MKLNdarray_set_structure(self, ndim, dims);
+    if (err < 0) {
+        return err;
```
Code was updated according to your comments.

pcs-theano commented on theano/contrib/mkl/mkl_ndarray.c:
```diff
@@ -0,0 +1,1361 @@
+    // prepare user layout and mkl buffer
+    err = MKLNdarray_create_buffer_from_structure(self);
+    if (err < 0) {
+        return err;
```
Code was updated according to your comments.

pcs-theano commented on theano/contrib/mkl/mkl_ndarray.c:
```diff
@@ -0,0 +1,1361 @@
+        return NULL;
+    }
+    // Fill with zeros
+    memset(rval->private_data, 0, total_size);
```
Yes, it's a potential issue, already fixed in the new codes.

pcs-theano commented on theano/contrib/mkl/mkl_ndarray.c:
```diff
@@ -0,0 +1,1361 @@
+        total_element = 0;
+    } else {
+        for (int i = 0; i < self->nd; i++) {
+            total_element *= self->user_structure[i];
```
Code was updated according to your comments.

pcs-theano commented on theano/contrib/mkl/mkl_ndarray.c:
```diff
@@ -0,0 +1,1361 @@
+                             "MKLNdarray_CreateArrayObj: dnnConversionCreate_F64 failed: %d, line: %d",
+                             status, __LINE__);
+                Py_DECREF(rval);
+                return NULL;
```
Code was updated according to your comments.

pcs-theano commented on theano/contrib/mkl/mkl_ndarray.c:
```diff
@@ -0,0 +1,1361 @@
+        status = dnnLayoutCreate_F64(&layout_user,
+                                     self->nd,
+                                     mkl_size,
+                                     mkl_stride);
```
Yes, you're right about the conversion. But we can't create the conversion primitive for two same layout, so we can't call the convert function (which requires conversion primitive as one of its parameters) to do the simple copy work.

No matter we convert private layout to user layout or vice verse, private layout is always an internal description for mkl object, but user layout is built from Numpy Ndarray actually.

For example, wehen convert private layout to user layout, so the destination layout is layout_user. Although the information we used to build the layout_user is `mkl_size` and `mkl_stride` (and they're stored in user_structure of mkl object), they're acquired from Numpy Ndarray (`dims` and `strides` respectively), the only difference between `mkl_size` and `dims` is element order, `mkl_size` use `WHCN` order, while `dims` use `NCHW` order, same for strides.

So mkl_strides should be ok in this case.

pcs-theano commented on theano/contrib/mkl/mkl_ndarray.c:
```diff
@@ -0,0 +1,1361 @@
+    PyObject* shape = NULL;
+    int typenum = -1;
+
+    if (!PyArg_ParseTuple(args, "O|i", &shape, &typenum)) {
```
Code was updated according to your comments.

pcs-theano commented on theano/contrib/mkl/mkl_ndarray.c:
```diff
@@ -0,0 +1,1361 @@
+    self->private_workspace = NULL;
+    self->data_size         = 0;
+    self->private_layout    = NULL;
+    memset((void*)(self->user_structure), 0, 2 * MAX_NDIM * sizeof (size_t));
```
Since we need strides information to build layout_user, so we need to keep this.

pcs-theano commented:
Hi @nouiz, any updates for this PR?
Currently, we're preparing other PRs such as convolution/pooling/relu, etc., but all these Ops will depend on this one since they're built based on mkl ndarray data type. If the current design of mkl ndarray is ok, then we can move forward with other Ops smoothly, or we need to make relevant changes according to new comments on mkl ndarray.

nouiz commented on theano/contrib/mkl/mkl_ndarray.h:
```diff
@@ -54,16 +59,18 @@ char* MKL_TYPE[] = {"", "", "", "int16", "", "int32", "", "int64",
+    size_t      user_structure[2 * MNDA_MAX_NDIM];  // user layout: [size0, size1, ..., stride0, stride1, ..., 0, 0].
+    dnnLayout_t private_layout;                     // layout instance create by MKL APIs
+    void*       private_data;                       // data buffer
+    dnnLayout_t private_layout_ws;                  // layout for workspace
```
Can you confirmt this extra layout is needed as sometimes the data layout and the workspace layout are different?

Currently, theano handle different workspace, but the way you do it is hidded from Theano. So it is fine. Is it possible that different node in the graph share the same workspace to lower the memory usage? I don't think it is mandatory, but if it is possible/already done, it would be great.

pcs-theano replied:
`private_layout_ws` is different than `private_layout`, workspace layout/buffer is needed to store the intermediate result by some MKL Ops such as pool/lrn. For MKL Op, only backward and forward node share the same workspace (for example, pool and poolGrad), not for different node.
So it's necessary to keep them.

pcs-theano commented on theano/contrib/mkl/mkl_ndarray.c:
```diff
@@ -6,95 +6,129 @@
+        rval = 0;
+
+    } else {
+        if (MNDA_FLOAT64 == self->dtype) {  // for float64
```
MNDA stands for MKL NDArray to avoid naming conflict with MKL library.

pcs-theano commented on theano/contrib/mkl/mkl_ndarray.c:
```diff
@@ -1139,7 +1540,23 @@ static PyMethodDef MKLNdarray_methods[] = {
+        "Create a new MklNdarray with specified shape, filled ith zeros."},
+
+    {"__copy__",
+        (PyCFunction)MKLNdarray_View, METH_NOARGS,
```
Sure, will align to latest codes.

pcs-theano replied:
Before submit the new code, did you mean both `__copy__` and `__deepcopy__` should do the same deepcopy, and keep the current `copy` and `view` as it is?

nouiz replied:
Change MKLNdarray_View to MKLNdarray_DeepCopy here.

nouiz commented:
Ok. Put that in documentation/comments somewhere.

But for the dtype, why not reuse MKL dtype? You don't need to change it, I
was curious why you redefined them instead of using there definition in
some MKL headers.

On Fri, Jun 23, 2017 at 2:22 AM pcs-theano <notifications@github.com> wrote:

> *@pcs-theano* commented on this pull request.
> ------------------------------
>
> In theano/contrib/mkl/mkl_ndarray.c
> <https://github.com/Theano/Theano/pull/5770#discussion_r123679408>:
>
> > -            if (rval != E_SUCCESS) {
> -                PyErr_Format(PyExc_RuntimeError,
> -                             "MKLNdarray_uninit: fail to release data: %d, line: %d",
> -                             rval, __LINE__);
> +    // self is a view
> +    if (NULL != self->base) {
> +        self->data_size = 0;
> +        self->workspace_size = 0;
> +        self->nd = -1;
> +        self->dtype = -1;
> +        Py_DECREF(self->base);
> +        self->base = NULL;
> +        rval = 0;
> +
> +    } else {
> +        if (MNDA_FLOAT64 == self->dtype) {  // for float64
>
> MNDA stands for MKL NDArray to avoid naming conflict with MKL library.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/5770#discussion_r123679408>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-1nmBEhlWItjztbZuo7Cl89EC74gks5sG1nAgaJpZM4Mp_44>
> .
>

pcs-theano commented:
The reason we changed to `MNDA_FLOATx` is: `MNDA_FLOATx` stands for type 
number (11, 12) to align with Numpy, which has different meaning compared to `MKL_INTx`
used by MKL library. For example `MKL_INT8` and `MKL_INT64` stands for
`int` or `long long int` types respectively, besides, there's no `MKL_FLOATx` in MKL.
Since MKLNdarray is not a part of MKL library, so we changed it to avoid namespace conflict
with MKL.
> But for the dtype, why not reuse MKL dtype? You don't need to change it, I
was curious why you redefined them instead of using there definition in
some MKL headers.
